src/lib/math/pcurves/pcurves.cpp

```diff
@@ -6,178 +6,177 @@
 
 #include <botan/internal/pcurves.h>
 
-#include <botan/internal/pcurves_wrap.h>
+#include <botan/internal/pcurves_instance.h>
 
-namespace Botan::PCurve {
-
-namespace {
-
-// clang-format off
-
-namespace secp256r1 {
-
-class Params final : public EllipticCurveParameters<
-   "FFFFFFFF00000001000000000000000000000000FFFFFFFFFFFFFFFFFFFFFFFF",
-   "FFFFFFFF00000001000000000000000000000000FFFFFFFFFFFFFFFFFFFFFFFC",
-   "5AC635D8AA3A93E7B3EBBD55769886BC651D06B0CC53B0F63BCE3C3E27D2604B",
-   "FFFFFFFF00000000FFFFFFFFFFFFFFFFBCE6FAADA7179E84F3B9CAC2FC632551",
-   "6B17D1F2E12C4247F8BCE6E563A440F277037D812DEB33A0F4A13945D898C296",
-   "4FE342E2FE1A7F9B8EE7EB4A7C0F9E162BCE33576B315ECECBB6406837BF51F5",
-   -10> {
-};
-
-class Curve final : public EllipticCurve<Params> {};
-
-}  // namespace secp256r1
+#if defined(BOTAN_HAS_ASN1)
+   #include <botan/asn1_obj.h>
+#endif
 
-namespace secp384r1 {
-
-class Params final : public EllipticCurveParameters<
-   "FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFFFF0000000000000000FFFFFFFF",
-   "FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFFFF0000000000000000FFFFFFFC",
-   "B3312FA7E23EE7E4988E056BE3F82D19181D9C6EFE8141120314088F5013875AC656398D8A2ED19D2A85C8EDD3EC2AEF",
-   "FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFC7634D81F4372DDF581A0DB248B0A77AECEC196ACCC52973",
-   "AA87CA22BE8B05378EB1C71EF320AD746E1D3B628BA79B9859F741E082542A385502F25DBF55296C3A545E3872760AB7",
-   "3617DE4A96262C6F5D9E98BF9292DC29F8F41DBD289A147CE9DA3113B5F0B8C00A60B1CE1D7E819D7A431D7C90EA0E5F",
-   -12> {
-};
-
-class Curve final : public EllipticCurve<Params> {};
+namespace Botan::PCurve {
 
+#if !defined(BOTAN_HAS_PCURVES_SECP256R1)
+//static
+std::shared_ptr<const PrimeOrderCurve> PCurveInstance::secp256r1() {
+   return nullptr;
 }
```

Comment:
The entire pcurves library is meant for internal use only, right? Otherwise, I'd really prefer a `std::optional<>` here. Missing the `nullptr` check is an easy thing to happen in downstream code. And we might even miss it in CI, if we neglect testing with a restricted build configuration. (That said: The BSI policy might be a good candidate to test this.)

Reply:
Right, this is all completely for internal use with zero expectation we'll ever expose this to library users.

What happens is users (and really, even almost all uses within the library) use `EC_Group` plus `EC_Scalar` and `EC_AffinePoint` and if the right curve is in use things go fast because we divert from the current EC code to here. The main place I can see where we *might* use this interface directly if if we were implementing some new algorithm (like say FROST) and did not expect or desire it to support arbitrary curves.

Regarding `optional<shared_ptr<T>>` IMO then you just have to remember to check twice :) once for the `optional` and then the `shared_ptr`